### ISSUE TITLE
Continious docker image vulnerability scanning

### ISSUE THREAD

**Issue by MOZGIII** [timberio/vector#5430]
To provide better security guarantees and to able to meet the industry standards on containerised application security, we have to implement a continious docker image vulnerability scanning.

## Motivation

The goals of this process are to guarantee that our we can we aware of and adeqately react to common security issues.

In some environments, having the images scanned for vulnerabilities is a mandatory step before the deployment, and we want Vector to be usable in those environments, by proactively resolving all issues that are discovered.

## Design considerations

One important spect is that we want to rescan all of our images all the time - on the contrary to scanning just the last or currently-build image at build time. The reason for it is that vulnerabilties are discovered and databases get updated over time, and we want to be able to detect the latest vulnerabilities even in the oldest images. Obtaining and properly and transparently presenting this information to our users is crucial, and can be decoupled from the process of resolving those issues (i.e. we might want to only ship security updates to just the images for the a couple latest Vector versions).
 
## How

This is to be determined currently.

**Comment by binarylogic** [comment]
cc @jamtur01 and @Hoverbear. We also discussed dropping support for distro-based images, which I assume would simplify this.

**Comment by binarylogic** [comment]
Adding another note that we'll want to add details of this process to the SECURITY.md file.

**Comment by MOZGIII** [comment]
> We also discussed dropping support for distro-based images, which I assume would simplify this.

Distro/distroless images should be no different in this aspect.

> Adding another note that we'll want to add details of this process to the SECURITY.md file.

I agree, although the process of scanning should be fully automatic, the resolution process for the detected vulnerabilities should be documented.

---

To start the discussion with something, how do you feel about using this: https://docs.docker.com/docker-hub/vulnerability-scanning/ ?

**Comment by binarylogic** [comment]
Docker hub works assuming we are also building our images there. In general, it would  be nice to consolidate on Github Actions, but I'm not too worried about it.

**Comment by MOZGIII** [comment]
I think it makes sense to offload this to an external provider either way, rather than build it at our own CI. It might make sense to search at the Github Marketplace if we like go this route.
There is a number of open-source tools there too, but integrating them properly might be non-trivial.

**Comment by Hoverbear** [comment]
Our dockerfiles take known base images and add binaries to them. https://github.com/timberio/vector/blob/master/distribution/docker/alpine/Dockerfile

I presume that one of these security flaggings would require us to bump our base image? Is there a reason for us not to use the latest baseimages for release? Could we run this check after our image builds maybe, to avoid doing much extra work?

**Comment by MOZGIII** [comment]
> I presume that one of these security flaggings would require us to bump our base image?

Speaking of steps to resolve the vulnerabilities - as lost as the vulnerability has come from the base image, and is fixed in the base image, we can simply pull the latest version and rebuild our image. No need to bump the image explicitly (as in changing the tag).

In some cases though, the vulnerability might be in our dependency code - this may or may not be detected at the image level depending on the details of the situation, but in such cases image rebuild is not enough, and we'll have to rebuild the binaries as well.

In some cases, the vulnerability may come from the build environment (the environment we use to produce the Vector binaries). In this case we'd need to apply necessary updates to the build environment, rebuild the binaries, and then rebuild the images.

> Is there a reason for us not to use the latest baseimages for release?

No reason, the latest base images are usually the most updated/fixed ones.

> Could we run this check after our image builds maybe, to avoid doing much extra work?

We need to run the check regularly for already built images that we intend to offer security updates for. The reason is that usually the vulnerabilities are discovered after the image is built/shipped, so our job would be to watch our artifacts closely and ship security updates (typically just simple rebuilds, but occasionally non-trivial fixes) - as needed. Doing this once at build time is unfortunately not enough, as we need to monitor for the newly discovered vulnerabilities as they are discovered.

**Comment by Hoverbear** [comment]
> In some cases though, the vulnerability might be in our dependency code - this may or may not be detected at the image level depending on the details of the situation, but in such cases image rebuild is not enough, and we'll have to rebuild the binaries as well.

We only depend on glibc, and only for glibc builds. It's always system provided, and we cannot safely patch it.

> In some cases, the vulnerability may come from the build environment (the environment we use to produce the Vector binaries). In this case we'd need to apply necessary updates to the build environment, rebuild the binaries, and then rebuild the images.

We do not build with these docker images, we load the binary artifacts. The images we use are official builders provided by upstream Cross project, and we link to static libraries in all cases.

**Comment by Hoverbear** [comment]
> We need to run the check regularly for already built images that we intend to offer security updates for. The reason is that usually the vulnerabilities are discovered after the image is built/shipped, so our job would be to watch our artifacts closely and ship security updates (typically just simple rebuilds, but occasionally non-trivial fixes) - as needed. Doing this once at build time is unfortunately not enough, as we need to monitor for the newly discovered vulnerabilities in our past artifacts as they are discovered.
>
> As an extension beyond the standard procedures of doing regular checks, we can also run the scan for each of the images we build before we push them - but that will be doing the extra work, not avoiding it, as it does not replace the regular checks procedure. What it will do for us though, is it allow us to know in time that the image we've just built has known security issues, and so we'll be able to react swiftly. But we still need to rescan our images constantly with an updated vulnerability database.


Yes!!! I feel like we can do this in an entirely automated manner maybe? I know Docker hub let us trigger on image updates of the base image so we could force a rebuild when that happens?

**Comment by MOZGIII** [comment]
> I feel like we can do this in an entirely automated manner maybe? I know Docker hub let us trigger on image updates of the base image so we could force a rebuild when that happens?

They used to have this, but I think it was disabled in a couple of months ago.

Entirely automating this is not possible, because fixing vulnerable things can sometimes be a manual process. But we definitely can (must!) script things like old image rebuilds.

However, I'm not aiming to address this automation initially - it can take time, and right now we need some reliable signals so we can react (manually for now) when we have the vulnerability. How exactly to react is an RFC-worthy process, and it's better we should gain some knowledge fixing things manually (or with simple throwaway automation) first.

**Comment by binarylogic** [comment]
@MOZGIII does reducing our distro support help with this? Ex: only offering distroless variants.

**Comment by MOZGIII** [comment]
I don't think so. I mean it does reduce the footprint, but for a process like this, the footprint isn't the issue. There's no difference in the kinds of images - they're all handed universally.

I'd be very hesitant with cutting distro-based images since the distress is not the same as "from scratch" images, and thus they have their own code that they bring along - and for that matter, distro-based images are better, because they have well-established vulnerability resolution pipelines, and many companies sec teams have already audited and accepted the bases into their clusters, unlike less popular distroless images.

**Comment by jamtur01** [comment]
+1 to not cutting distro-based images.

**Comment by jamtur01** [comment]
And yes - this definitely merits an RFC (albeit a brief one) that explains:

1. What we want to achieve.
2. What we want to look for
3. Our SLA/policy for fixes (and what we do when upstream hasn't released a fix).
4. How far back any potential rebuilds go (I am definitely worried about changing the underlying image and introducing new bugs whilst fixing a security issue).

etc...

**Comment by MOZGIII** [comment]
One of the possibilities we could review to have complete control over our artifacts is completely vendoring our toolchains (clang, rustc, etc) - somewhat like embedded projects do. Counterintuitively, this should be easier than trying to wire our code via some distro build system, and more reliable in the long run. However, it's just like with rust - with a greater degree of control you inevitably get increased complexity. Albeit, the complexity there will go in the right place - actually maintaining the dependencies and the toolchain directly, rather than inventing hacks to wire things through different existing toolchains. This has some serious benefits - i.e. no distro-specific patches to headers files and libraries we don't see will affect Vector artifacts. However it a lot of work.

In the context that we've just made things work for us via cross-build, I can definitely understand people want some peacetime in this regard, so I don't think it would be wise to move beyond just the discussion of this topic for now.

**Comment by MOZGIII** [comment]
> @Hoverbear's point that we add **binaries** to Docker images - we don't build inside these images - is also important to factor in as a consideration.

Absolutely!

With the resulting image scanning approach, it shouldn't really matter how we produce the images. I think I've seen scanners properly detect that we have a lua vulnerability, even though it's statically linked - right from our single binary. So I think they do something intelligent there, i.e. split the binaries into sections and actually scan the executable code.

**Comment by MOZGIII** [comment]
I propose that we address this in the following steps:
1. Quickly setup some image scanning and vulnerability notification channels, just for now - the goal is to start getting the data points as quickly as possible - tracked by this issue.
1.1. Discuss and fix things that we uncover immediately.
2. Write an RFC about scanning and the vulnerability resolution process.
3. Implement scanning and resolution protocol according to the RFC.

**Comment by jamtur01** [comment]
Then I vote we upgrade our Docker Hub account and start scanning there. Since that's all in the once place.

**Comment by MOZGIII** [comment]
> Then I vote we upgrade our Docker Hub account and start scanning there. Since that's all in the once place.

Sounds good! It'll immediately solve our problem for the short term - just what we want.

**Comment by MOZGIII** [comment]
Looks like Docker Hub and other docker tooling use https://snyk.io under the hood. We might want to integrate with them directly instead - might be quicker than setting up upgraded billing for Docker Hub.

**Comment by jszwedko** [comment]
I stumbled across https://github.com/quay/clair today too, as another option.

**Comment by MOZGIII** [comment]
Yep, there's a number of those, and there's even `docker scan <image>` built-in.
